### PR TITLE
fix: custom menu for guide

### LINKED ISSUE #35287: The guide's mobile menu button should open guides side nav instead of showing main nav items
The guides does not have the way to navigate through the articles. we would like to have the top menu button on the nav to go display the side nav instead of showing the search bar and other main nav items.

here is what happens when it is the menu button is clicked
screen shot 2019-02-19 at 8 33 24 am

here is what should happen.
screen shot 2019-02-19 at 8 33 01 am

### PR DESCRIPTION
**Pull request by ahmaxed**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #35287

### CODE REVIEW

**Comment by ahmadabdolsaheb** [comment]
This PR will make a custom menu tab for the guides on mobile.
features:
- change the main navigation's breakpoint to occur when the side nav disappears.
- disable expanding the mobile navigation on guide and display side nav.
- added styles to the menu button to differentiate active state.
- removed GuideNavigationContext.js and relocated its state to redux store.
- opening and closing of menu button is wired to redux so the menu button could be closed when an end node on the side nav
-close the nav menu when the guide is unmounted so the menu would not be open when other pages are visited.

edge-case:
since the logic to render nav items takes place when header is mounted, the nav buttons doesn't appear when the menu is open on guide and the screen's width is increased(resized). the solution would be to add an event listener to and change the nav if the width of the screen is changed, but it seems to be wasteful.

**Comment by raisedadead** [comment]
Rebase for a QA?

**Review by raisedadead — APPROVED** [review]
Functionality LGTM. I'll let others confirm on the code.

**Comment by raisedadead** [comment]
@ahmadabdolsaheb while you are at this can you see why we have a jumpy render when browsing in desktop mode.

I think the bread crumb is not rendered before the content or something.

**Comment by raisedadead** [comment]
I have removed the critical path. We are keeping it for absolute show stoppers. Need one more rebase, sorry :(

**Comment by ahmadabdolsaheb** [comment]
> @ahmadabdolsaheb while you are at this can you see why we have a jumpy render when browsing in desktop mode.
> 
> I think the bread crumb is not rendered before the content or something.

You mean the tiny refresh that happens after click? it only happens on some nodes interestingly. yes sure as soon as we are done with this one.

**Comment by ahmadabdolsaheb** [comment]
@ValeraS pointed out this is causing the problem @raisedadead mentioned above. I was wondering if we could go ahead and remove it ,@Bouncey 
https://github.com/freeCodeCamp/freeCodeCamp/blob/dd9ce2bc2aec657e915bb7bef2b938db6a721e37/client/src/components/layouts/Guide.js#L35

File: client/src/components/Header/index.js [hunk 1]
```diff
@@ -9,12 +11,31 @@ import { Link } from '../helpers';
...
 import './header.css';
 
+import {
+  toggleDisplaySideNav,
+  toggleDisplayMenu
+} from '../layouts/components/guide/redux';
+
+const mapStateToProps = state => {
+  return {
+    displayMenu: state.guideNav.displayMenu
```
Comment (Bouncey) [inline client/src/components/Header/index.js]: The selector pattern seeks to decouple state from consumers.

Currently, if we changed the shape of the application state, we would have to visit these files so they still received the correct state. For example, if I changed the `guideNav` namespace.

If we used a selector, we could restructure the state and only have to worry about how the selector accessed the state required. The consumers would not know that the shape of state has changed.

**TLDR;** State consumers should not care for the shape of the state they are consuming, selectors to the rescue! 💪

File: client/src/components/Header/index.js [hunk 2]
```diff
@@ -24,59 +45,85 @@ class Header extends Component {
 
   componentWillUnmount() {
     document.removeEventListener('click', this.handleClickOutside);
+    console.log('header unmount');
```
Comment (Bouncey) [inline client/src/components/Header/index.js]: `console.log`

File: client/src/components/Header/index.js [hunk 3]
```diff
@@ -24,59 +45,85 @@ class Header extends Component {
...
-            <li>
-              <Link external={true} to='/news'>
-                News
-              </Link>
-            </li>
-            <li className='user-state-link'>
-              <UserState disableSettings={disableSettings} />
-            </li>
-          </ul>
+          {onGuide && displayMenu && viewportWidth < 991 ? null : (
```
Comment (Bouncey) [inline client/src/components/Header/index.js]: We use `react-media` for these types of media queries

File: client/src/components/layouts/Guide.js [hunk 4]
```diff
@@ -26,9 +32,25 @@ const propTypes = {
...
+  displaySideNav: PropTypes.bool,
+  expandedState: PropTypes.object,
+  location: PropTypes.object,
+  toggleDisplaySideNav: PropTypes.func,
+  toggleExpandedState: PropTypes.func
+};
+
+const mapStateToProps = state => {
+  return {
+    expandedState: state.guideNav.expandedState,
```
Comment (Bouncey) [inline client/src/components/layouts/Guide.js]: See comment about selectors

File: client/src/components/layouts/components/guide/SideNav.js [hunk 5]
```diff
@@ -27,6 +27,30 @@ class SideNav extends Component {
     this.renderParent = this.renderParent.bind(this);
   }
 
+  componentDidMount() {
```
Comment (Bouncey) [inline client/src/components/layouts/components/guide/SideNav.js]: This logic would be better handled in a saga, much like we handle the `challengeMounted` logic.

**Review by Bouncey — CHANGES_REQUESTED** [review]

File: client/src/components/layouts/components/guide/SideNav.js (hunk 5 above)
Comment (ahmadabdolsaheb) [reply client/src/components/layouts/components/guide/SideNav.js]: @Bouncey so I am trying to find the most efficient way to execute this in saga, here is what I got so far:
the saga will listen for app-mount action, and check to see if the user is in the guide by checking the window.location.pathname. if so, it will dispatch toggleExpandedState accordingly.
Since this is going to be called on every app-mount, do you think this is a good approach?

**Review by Bouncey — APPROVED** [review]

File: client/src/components/layouts/components/guide/redux/side-navigation-saga.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { takeEvery, put, all } from 'redux-saga/effects';
+
+import { toggleExpandedState } from './';
+import { types as appTypes } from '../../../../../redux';
+
+function* showDonateModalSaga() {
```
Comment (Bouncey) [inline client/src/components/layouts/components/guide/redux/side-navigation-saga.js]:
```suggestion
function* expandSideNavSaga() {
```

File: client/src/components/layouts/components/guide/redux/side-navigation-saga.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      }, {});
+
+    let routes = Object.keys(pathMap).map(key => pathMap[key]);
+
+    yield all(routes.map(route => put(toggleExpandedState(route))));
+  }
+}
+
+export function createSideNavigationSaga() {
+  return [takeEvery(appTypes.appMount, showDonateModalSaga)];
```
Comment (Bouncey) [inline client/src/components/layouts/components/guide/redux/side-navigation-saga.js]:
```suggestion
  return [takeEvery(appTypes.appMount, expandSideNavSaga)];
```

**Review by Bouncey — CHANGES_REQUESTED** [review]

**Review by Bouncey — APPROVED** [review]
